### ISSUE TITLE
Add ignoreContextFunctionalPseudoClasses: [] to selector-max-id

### ISSUE THREAD

**Issue by superfein** [stylelint/stylelint#4619]
I have multiple instances of a :not() CSS pseudo class selector with an ID as a parameter, causing errors in the console. Multiple instances once the SCSS is processed to CSS.

> Which rule, if any, is the bug related to?

e.g. `selector-max-id`

> What code is needed to reproduce the bug?

```
.comment-form {
	width: 100%;

	input:not(#wp-comment-cookies-consent):not(.button) {
		clear: both;
		transition: $transition-border;
		width: 100%;

		&:focus,
		&:active {
			border: 1px solid $color-border-comment-focus;
		}
        }
}
```

> Which version of stylelint are you using?

`13.1.0`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to SCSS nested properties.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged: 
```
 59:12  ×  Expected "#wp-comment-cookies-consent" to have no more than 0 ID selectors   selector-max-id
 65:3   ×  Expected "#wp-comment-cookies-consent" to have no more than 0 ID selectors   selector-max-id
 65:3   ×  Expected "#wp-comment-cookies-consent" to have no more than 0 ID selectors   selector-max-id
```

**Comment by jeddy3** [comment]
@superfein Thanks for the report and for using the template.

I think the rule is behaving as expected. There is an id selector within the `:not` functional pseudo-class.

You can either:

- change your configuration to allow up to one id per selector
- change your selector to not use an id

However, if neither of these are appropriate because you'd like to _allow id selectors within `:not` functional pseudo-classes and disallow id selectors everywhere else_ then I think we can add an `ignoreFunction[]` optional secondary object to make the rule more permissive.

For example:

```json
{
  "rules" : { "selector-max-id": [ 0, { "ignoreFunctions": ["not"] } ] }
}
```

Do any of those three suggestions resolve the issue for you?

**Comment by superfein** [comment]
Hi @jeddy3, thanks for your response! I would like to see the third option, an optional secondary object to to make the rule more permissive. I think that will increase specificity and allow for greater control. When working with a CMS sometimes targeting a specific element can come down to an ID selector, so adding the secondary object to the `ignoreFunction[]` would be a huge asset.

**Comment by malsf21** [comment]
I can probably take this if nobody else is working on it!

**Comment by malsf21** [comment]
Just working away at this issue, but just want to clarify the requested functionality: by `ignoreFunctions`, should I be looking for both function nodes (like `foo(#bar)`, which already is not affected by the rule) and pseudo-function nodes (like `input:not(#bar)`)? I'm also slightly confused by the type of the latter, will that nodes' type be `pseudo`?

**Comment by jeddy3** [comment]
> should I be looking for both function nodes (like `foo(#bar)`, which already is not affected by the rule

`foo(#bar)` is not standard CSS construct and the rule correctly ignores any selectors that contain it.

> pseudo-function nodes (like input:not(#bar))?

Yes, specifically [_functional pseudo-classes_](https://drafts.csswg.org/selectors-4/#pseudo-classes) that provide [_evaluation contexts_](https://drafts.csswg.org/selectors-4/#specificity-rules). I suggest we name the option `ignoreContextFunctionalPseudoClasses: []` to reflect this. It's verbose, but it accurately reflects what it does.

In the rule, we can update the following conditional:

https://github.com/stylelint/stylelint/blob/4a4ea84c22145e7ae5d3eece3d8cf1afd7cec6d8/lib/rules/selector-max-id/index.js#L38

To

```js
if (
  childNode.type === 'selector' || 
  (isLogicalCombination(childNode) && !isIgnoredContextFunctionalPseudoClass(childNode, options))
) {
```

Where `ignoredContextFunctionalPseudoClass` looks something like:

```js
function isIgnoredContextFunctionalPseudoClass(node, options) {
  if (node.type === 'pseudo' && optionsMatches(options, 'ignoreContextFunctionalPseudoClass', node.value) return true
  else return false
}
```

The new option will allow users to selectively ignore _functional pseudo-classes which provide evaluation contexts_ that the rule currently evaluates separately.

**Comment by malsf21** [comment]
Wow, thank you @jeddy3 for the in-depth help! I should be able to implement this after work. Really appreciate it!

**Comment by malsf21** [comment]
Hi @jeddy3, as I was implementing the functionality (which I've completed in a fork), I've come across either a point of confusion or an inconsistency with how `selector-max-id` works on other evaluation context providing functional pseudo-classes like `:where` and `:is`.

For example, examine [this demo](https://stylelint.io/demo#N4Igxg9gJgpiBcICG8DuALGAnGAKAxAGYQQCUwkANhFvAAQBGlSYA1gL4DcAOgHZ8oAlgGdcMALYAaOkRLkqNekxYdOISSGEBPXgBckADwThhw9eAi9CggObHgfOnW4gsAV0oxhL+g95OnF2EYTzBdGgBacUMIwSgXOnoABkc6dj52EHYgA):

```css
input:where(#foo){
  color: black
};

a:is(em, #foo){
  color: black
};
```

seems like it should cause an error with the following config 

```json
{
  "rules": {
    "selector-max-id" : 0
  }
}
```

but it doesn't. Am I using the rule properly? I'm a bit confused. It does work as intended for `:not` and `:has`.

**Comment by jeddy3** [comment]
> Am I using the rule properly? I'm a bit confused. It does work as intended for :not and :has.

Yes, you are using it currently. The rule doesn't support `:is` and `:where` yet. See [this demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIDEAzCEAEwC+AdAdgIbz4QAuAFKugJRZ5HwAWhAzlWhHTgcQLaEyYJjHbUu9Ar3gB3EQCcYHWpMYBLdjD4AaDOO4MQ2kKwCe+MoQAeCcK1ZHwEfCjUBzW8AIYMuEPIBXABtRP3gsbx9fExgQsDIIeQBaASsktSgwjAAGSIZsEGwgA). They are new pseudo-classes, e.g. `:is` is the new name for `:matches`, which in turn was called `:any`. 

It's a separate issue to update the underlying utility to support them and address these false negatives:

https://github.com/stylelint/stylelint/blob/58aab8def87fdceec4bbc4e87591892def144025/lib/utils/isLogicalCombination.js#L9-L22

We can make this new option work for just `:has`, `:not` and `:matches` for now. I suggest using `:not` in your tests and README example as it's the most common of the lot.

**Comment by malsf21** [comment]
Great, just submitted the PR. Would you like me to create another issue and PR to update the `isLogicalCombination` utility?

**Comment by jeddy3** [comment]
> Would you like me to create another issue and PR to update the `isLogicalCombination` utility?

That'll be great!
